lib/push_2_tweet_communication.js

```diff
@@ -0,0 +1,156 @@
+ * published by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * push-2-tweet is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public
+ * License along with push-2-tweet.
+ * If not, seehttp://www.gnu.org/licenses/.
+ *
+ * For those usages not covered by the GNU Affero General Public License
+ * please contact with::[german.torodelvalle@telefonica.com]
+ */
+
+'use strict';
+
+var p2tConfig = require('./push_2_tweet_configuration'),
+  p2tLogger = require('logops'),
+  p2tHelper = require('./push_2_tweet_helper'),
+  caErrors = require('./push_2_tweet_error'),
+  sendRequest = require('request'),
+  revalidator = require('revalidator'),
+  payloadJSONSchema = require('./schema/request-payload-schema.json');
+
+/**
+ * Checks that a request sent to the Push 2 Tweet is well-formed
+ * @param {object} request The request sent to the Push 2 Tweet
+ * @param {function} callback Callback to be called with the result of the check
+ * @return {*} boolean True if the updateContext request is valid. False otherwise
+ */
+function checkRequest(request, callback) {
+  var payloadJSON;
+
+  if (typeof(request.payload) === 'string') {
+    try {
+      payloadJSON = JSON.parse(request.payload);
+    } catch (err) {
+      p2tLogger.warn(request.push2Tweet.context, 'Error when parsing the request payload: ' + request.payload +
+        (err.message ? ', error=' + err : '')
+      );
+      return process.nextTick(callback.bind(null, err));
+    }
+  } else {
+    payloadJSON = request.payload;
+  }
+  var validation = revalidator.validate(payloadJSON, payloadJSONSchema);
+  if (validation.valid) {
+    p2tLogger.debug(request.push2Tweet.context, 'The request payload: ' + request.payload + ' is valid');
+    return process.nextTick(callback);
+  } else {
+    var err = new caErrors.BadPayload(
+      request.payload
+    );
+    err.message += ' (errors: ' + p2tHelper.stringifyRevalidatorErrors(validation.errors) + ')';
+    p2tLogger.warn(request.push2Tweet.context, 'The request payload: ' + request.payload + ' is NOT valid' +
+      (err.message ? ', error=' + JSON.stringify(err) : '')
+    );
+    return process.nextTick(callback.bind(null, err));
+  }
+}
+
+/**
+ * Processes and returns the requested operation information
+ * @param {object} request The updateContext request received
+ * @param {function} callback Callback to be called with the result of the operation
+ */
+function getOperationDescriptor(request, callback) {
+  checkRequest(request, function(err) {
+    if (err) {
+      return process.nextTick(callback.bind(null, err));
+    } else {
+      var payloadJSON = JSON.parse(request.payload);
```

Comment:
It would be easier if you sepparate concerns in the `checkRequest()` function. Maybe you can separate into a parsing function and a validate function, and use the result of the former both for the validation and the extraction of the operation description.